.github/workflows/release.yml

```diff
@@ -0,0 +1,107 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+name: Release
+
+on:
+  push:
+    branches:
+      - "maint-[0-9]+.[0-9]+.[0-9]+"
+      - "release-[0-9]+.[0-9]+.[0-9]+"
+      - "release-[0-9]+.[0-9]+.[0-9]+-rc[0-9]+"
```

Comment:
Hmm. I think that we don't need run this workflow for branches. We need to run only for tags because this workflow creates a GitHub Release and it's associated with a Git tag.

@assignUser What do you think about this?

Reply:
The question is if we want to limit this to only respond to tags on these branches, which for the moment would be correct. Though it is rather unlikely that someone pushes a tag that matches the release tag to main or some other branch by accident... so :shrug: Better save than sorry? 

Btw I don't remember where the `release-*` branches come from? Are they created after the vote? What is their purpose compared to `maint-*`?

Reply:
I think that a Git tag doesn't associated any branch. So `apache-arrow-X.Y.Z-rcN` tag on the `main` branch nor `apache-arrow-X.Y.Z-rcN` tag on the `release-X-Y-Z` branch don't exist.

> Btw I don't remember where the `release-*` branches come from? Are they created after the vote? What is their purpose compared to `maint-*`?

I think:
* `release-*` are for preparing `apache-arrow-X.Y.Z` tags
* `maint-X.Y.x` are for preparing `release-X.Y.Z` branches

Reply:
I don't think we want to trigger these workflows for branches. However, I could add a conditional to prevent the `Publish` job from running if the branch name doesn't match `release-*` or `maint-*`.  How does that sound?

Reply:
I just realized I completely misread Kou's comment about git tag's not being associated specific branches. Sorry about that! You can disregard my previous comment.

Reply:
Based on the discussion, it seems like we should not trigger the release and release candidate workflows when pushing specific branches. Marking as resolved.